README.md
# randy
#### zulip-bot providing random number generation and game objects

This bot is designed to respond to requests for coin flips and dice rolls, currently it can respond to `coin` with a single coin flip and `dice` with a single D6 roll. It also has a help function on `help` and responds to any other kind of message with a generic humanesque response. A closer look at the code will tell you it is designed for much more than this however and there is much intention for improvement. The intent is to remove the parsing limitations and parse a variation of messages from very explicit to strealined with a great deal more commands included. Hopefully in time it will have functionality for, multiple requests, a choice of faces for the dice, weightings to create trick coins and loaded dice, simple mathematical functions on outputs, better IO handling, more descriptive help messaging, standard random number generation, and a card based system.


CHANGELOG
------------
Randy-Alpha-5F06-5E00-Patch
------------
Fixed bug - double increment at the bottom of die.py
Fixed bug - testing empty string not NoneType at line 88 of randy.py
Fixed bug - convert error to string first before piping to content
Ammended - removed from . imports to fall in line with new hosting method


------------
Randy-Alpha-5F05-0C80
------------
Created program complete with witty retorts, error handling and help function

Name: Randy
Synopsis:
{curly braces define optional input}
------------
Coin; @randy {<times> }c{oins weight <heads>:<tails>:<clumsy>}
Examples; long:- `@randy 20 coin weight 20:10:30`, short:- `@randy 20cw2:1:3`
------------
Dice; @randy {<times> }d{ice <faces>,<stop>,<step> weight <number> <load>:<rest>:<clumsy> <operand>}
Examples; long:- `@randy 20 dice 16,24,2 weight 16 20:10:30 +89`, short:- `@randy 20d16,24,2w16 20:10:30+89`
------------
Times - the number of times the action will be performed, with dice all results will be automatically accumulated and a final result given.
Coin - the command used to define a coin flip.
Dice - the command used to define a dice roll.
Faces - the number of faces you want the die to have, becomes dice start value when stop and step are defined, with a default of 1.
Stop - defines what number to stop at when deciding the size of the dice. Both start and stop are inclusive so put the number you would expect on the dice.
Step - defines what the increments on the dice should be.
Weight - the command given to inform that you wish to edit the weightings.
Number - the number on the dice you wish to load. This must be followed by a space and then the ratio else it will default to 2:1.
Heads - the weighting you wish heads to have. Weightings should be integer ratios, larger numbers mean it is more likely lower is less likely.
Load - the weighting you wish the load number to have.
Tails - the weighting you wish tails to have.
Rest - the weighting you wish the rest of the numbers to have.
Clumsy - this is a weighting with a default of 0 that allows the potential for the item to be lost during operation.
Operand - this defines an operand  `+`, `-`, `/`, or `*` that you can perform upon the accumulated dice rolls. This is mostly useful for applying buffs or debuffs.

test.py
#!/usr/bin.env python3


class test:
    def __init__(self, value):
        self.value = value
        return;

    def ifExists(self, default):
        if self.value is None:
            self.value = default
        else:
            self.value = int(self.value)
        return;


die.py
#!/usr/bin/env python3


import random
import test

class die:
    def __init__(self, start, stop, step):
      if stop < start:
          raise TypeError("stop less than start? what're you trying to do? make an anti-dice?! XD")
      if (stop - start) % step != 0:
          raise TypeError ("hey dufus! might wanna check your math, that ain't a valid dice.")
      self.faces = int((stop + 1 - start) / step)
      self.weights = {}
      for a in range(start, stop + 1, step):
        self.weights[a] = 1
      self.clumsy = 0
      return;

    def setWeight(self, face, weight):
      if face not in self.weights.keys():
          raise TypeError ("yo genius, that number isn't on the dice, get a clue!")
      self.weights[face] = weight
      return;

    def setClumsy(self, weight):
      self.clumsy = weight
      return;

    def __rollDie(self):
        n = 1
        for a in self.weights.values():
            n += a
        n += self.clumsy
        pseudodie = random.randrange(n)
        valueSum = 0
        result = "lost"
        for key, value in self.weights.items():
            valueSum += value
            if pseudodie <= valueSum:
                result = key
                break
        return result;

    def rollDice(self, quantity):
        dice = [self.__rollDie() for a in range(quantity)]
        return dice;

    def parseDice(self, times, start, number, load, rest, clumsy):
        times.ifExists(1)
        number.ifExists(start.value)
        load.ifExists(1)
        rest.ifExists(1)
        clumsy.ifExists(0)
        for face in self.weights.keys():
            self.setWeight(face, rest.value)
        self.setWeight(number.value, load.value)
        self.setClumsy(clumsy.value)
        results = self.rollDice(times.value)
        return results;

def createSimpleDie(sides):
    return die(1, sides, 1);


coin.py
#!/usr/bin/env python3


import random
from . import test

class coin:
    def __init__(self):
        self.weightHeads = 1
        self.weightTails = 1
        self.clumsy = 0
        return;

    def __setHeads(self, weight):
        self.weightHeads = weight
        return;

    def __setTails(self, weight):
        self.weightTails = weight
        return;

    def __setClumsy(self, weight):
        self.clumsy = weight
        return;

    def __flipCoin(self):
        pseudobool = random.randrange(self.weightHeads + self.weightTails + self.clumsy + 1)
        if pseudobool <= self.weightHeads:
            result = "heads"
        elif pseudobool <= self.weightHeads + self.weightTails:
            result = "tails"
        else:
            result = "lost"
        return result;

    def flipCoins(self, quantity):
        coins = [self.__flipCoin() for a in range(quantity)]
        return coins;

    def weightCoin(self, heads, tails, clumsy):
        self.__setHeads(heads)
        self.__setTails(tails)
        self.__setClumsy(clumsy)
        return;

    def parseCoin(self, times, heads, tails, clumsy):
        times.ifExists(1)
        heads.ifExists(1)
        tails.ifExists(1)
        clumsy.ifExists(0)
        self.weightCoin(heads.value, tails.value, clumsy.value)
        results = self.flipCoins(times.value)
        return results;


message.py
#!/usr/bin/env python3


import re
from . import test

class message(object):
    def __init__(self):
        self.__text = False
        self.__match_coin = False
        self.__match_die = False
        self.__matches = []
        return;

    def set_text(self, text):
        self.__text = (text.strip()).lower
        return;

    def get_request(self):
        if self.__matches = []:
            raise TypeError("set matches first!")
        return self.__matches;

    def validate(self):
        if self.__text == False:
            raise TypeError("set text first!")
        elif self.__text == "":
            raise TypeError("empty string!")
        self.__match_coin = re.search("(([0-9]+)? ?)c(oin(s)?)?( ?w(eight)? ?([0-9]+):([0-9]+)(:([0-9]+))?)?", self.__text)
        self.__match_dice = re.search("((([0-9]+)? ?)d(ice)?( ?([0-9]+)(,([0-9]+),([0-9]+))?)?( ?w(eight)? ?([0-9]+) ([0-9]+):([0-9]+)(:([0-9]+))?)?( ?(\+|\-|\*|\/) ?([0-9]+))?", self.__text)
        return;

    def set_matches(self):
        if self.__match_coin == False and self.__match_dice == False:
            raise TypeError("validate input first!")
        elif self.__match_coin != None:
            self.__matches[0] = match_coin.group(2)
            self.__matches[1] = match_coin.group(7)
            self.__matches[2] = match_coin.group(8)
            self.__matches[3] = match_coin.group(10)
        elif self.__match_die != None:
            self.__matches[0] = match_dice.group(2)
            self.__matches[1] = match_dice.group(5)
            self.__matches[2] = match_dice.group(7)
            self.__matches[3] = match_dice.group(8)
            self.__matches[4] = match_dice.group(11)
            self.__matches[5] = match_dice.group(12)
            self.__matches[6] = match_dice.group(13)
            self.__matches[7] = match_dice.group(15)
            self.__matches[8] = match_dice.group(17)
            self.__matches[9] = match_dice.group(18)
        else:
            raise TypeError("no match!")
        return;

def get_message(name, text):
    name = message()
    name.set_text(text)
    try:
        name.validate()
        name.set_matches()
        return name.get_request();
    except TypeError as error:
        print("Error: invalid input, {0}".format(error))

expression.py
#!/usr/bin/env python3


import re
import test

class expression(object):
    def __init__(self):
        self.__text = False
        self.__match_coin = False
        self.__match_dice = False
        self.__matches_coin = []
        self.__matches_dice =[]
        self.__matches = []
        return;

    def set_text(self, text):
        self.__text = (text.strip()).lower()
        return;

    def get_request(self):
        if self.__matches == []:
            raise TypeError("set matches first!")
        return self.__matches;

    def validate(self):
        if self.__text == False:
            raise TypeError("set text first!")
        elif self.__text == "":
            raise TypeError("empty string!")
        self.__match_coin = re.search("(([0-9]+) ?)?c(oin(s)?)?( ?w(eight)? ?([0-9]+):([0-9]+)(:([0-9]+))?)?", self.__text)
        self.__match_dice = re.search("(([0-9]+) ?)?d(ice)?( ?([0-9]+)(,([0-9]+),([0-9]+))?)?( ?w(eight)? ?([0-9]+) ([0-9]+):([0-9]+)(:([0-9]+))?)?( ?(\+|\-|\*|\/) ?([0-9]+))?", self.__text)
        return;

    def set_matches(self):
        if self.__match_coin == False or self.__match_dice == False:
            raise TypeError("validate input first!")
        if self.__match_coin is not None:
            iterator = 0
            for item in self.__match_coin.group(2, 7, 8, 10):
                self.__matches_coin.append(item)
                iterator += 1
        if self.__match_dice is not None:
            iterator = 0
            for item in self.__match_dice.group(2, 5, 7, 8, 11, 12, 13, 15, 17, 18):
                self.__matches_dice.append(item)
                iterator += 1
        if self.__match_coin is not None and self.__match_dice is not None:
            if len(self.__match_coin.group(0)) <= len(self.__match_dice.group(0)):
                self.__matches = self.__matches_dice
            else:
                self.__matches = self.__matches_coin
        elif self.__match_coin is None:
            self.__matches = self.__matches_dice
        elif self.__match_dice is None:
            self.__matches = self.__matches_coin
        else:
            self.__matches = None
        return;

    def get_message(self, text):
        self.set_text(text)
        try:
            self.validate()
            self.set_matches()
            return self.get_request();
        except TypeError as error:
            print("Error: invalid input, {0}".format(error))


randy.py
#!/usr/bin/env python3


import expression
import coin
import die
import test

class Randy(object):
    '''
    This bot generates random coins and dice with biases
    '''

    def usage(self):
        return '''This bot generates random coins and dice with biases'''

    def handle_message(self, message, bot_handler):
        content = False
        text = message['content']

        # detects help message
        if "help" in text:
            content =   """
Name: Randy
Synopsis:
{curly braces define optional input}
------------
Coin; @randy {<times> }c{oins weight <heads>:<tails>:<clumsy>}
Examples; long:- `@randy 20 coin weight 20:10:30`, short:- `@randy 20cw2:1:3`
------------
Dice; @randy {<times> }d{ice <faces>,<stop>,<step> weight <number> <load>:<rest>:<clumsy> <operand>}
Examples; long:- `@randy 20 dice 16,24,2 weight 16 20:10:30 +89`, short:- `@randy 20d16,24,2w16 20:10:30+89`
------------
Times - the number of times the action will be performed, with dice all results will be automatically accumulated and a final result given.
Coin - the command used to define a coin flip.
Dice - the command used to define a dice roll.
Faces - the number of faces you want the die to have, becomes dice start value when stop and step are defined, with a default of 1.
Stop - defines what number to stop at when deciding the size of the dice. Both start and stop are inclusive so put the number you would expect on the dice.
Step - defines what the increments on the dice should be.
Weight - the command given to inform that you wish to edit the weightings.
Number - the number on the dice you wish to load. This must be followed by a space and then the ratio else it will default to 2:1.
Heads - the weighting you wish heads to have. Weightings should be integer ratios, larger numbers mean it is more likely lower is less likely.
Load - the weighting you wish the load number to have.
Tails - the weighting you wish tails to have.
Rest - the weighting you wish the rest of the numbers to have.
Clumsy - this is a weighting with a default of 0 that allows the potential for the item to be lost during operation.
Operand - this defines an operand  `+`, `-`, `/`, or `*` that you can perform upon the accumulated dice rolls. This is mostly useful for applying buffs or debuffs.
"""

        # prints purpose
        elif "purpose" in text:
            content = "My purpose is as a random item generator, I provide coin flips and dice rolls upon request, try `help` to to learn my commands."

        # responds to null input so you can check it's listening
        elif text == "" or text == "?":
            content = "yes?"
        
        else:
            command = expression.expression()
            matches = command.get_message(text)
            if matches == None:
                matches = []

            # detects coin input
            if len(matches) == 4:
                times = test.test(matches[0])
                heads = test.test(matches[1])
                tails = test.test(matches[2])
                clumsy = test.test(matches[3])
                operand = ""
                item = coin.coin()
                results = item.parseCoin(times, heads, tails, clumsy)
                slug = "You flipped a coin " + str(times.value) +  " times, the result was:\n"

            # detects dice input
            elif len(matches) == 10:
                times = test.test(matches[0])
                faces = test.test(matches[1])
                stop = test.test(matches[2])
                step = test.test(matches[3])
                number = test.test(matches[4])
                load = test.test(matches[5])
                rest = test.test(matches[6])
                clumsy = test.test(matches[7])
                operand = matches[8]
                value = test.test(matches[9])
                try:
                    if stop.value is None:
                        faces.ifExists(6)
                        item = die.createSimpleDie(faces.value)
                    else:
                        faces.ifExists(1)
                        stop.ifExists(6)
                        step.ifExists(1)
                        item = die.die(faces.value, stop.value, step.value)
                except TypeError as error:
                    content = str(error)
                else:
                    try:
                        results = item.parseDice(times, faces, number, load, rest, clumsy)
                    except TypeError as error:
                        content = str(error)
                    else:
                        slug = "You rolled a D" + str(len(item.weights.keys())) + " " + str(times.value) + " times, the result was:\n"

            # catches any other kind of input
            else:
                content = "I'm sorry what was that? I don't understand.\nType `help` to see the commands I know."

            # parses output in message friendly way
            if content == False:
                content = ""
                acc = 0
                for item in results:
                    if isinstance(item, int):
                        acc += item
                    content += ", " + str(item)
                content = content[2:]
                content = slug + content
                if operand is None and len(matches) == 10:
                    content += " = " + str(acc)
                elif operand is not None and len(operand) != 0:
                    value.ifExists(1)
                    try:
                        if operand == "+":
                            acc += value.value
                        elif operand == "-":
                            acc -= value.value
                        elif operand == "*":
                            acc *= value.value
                        elif operand == "/":
                            acc /= value.value
                    except ZeroDivisionError:
                            content = "divide by zero? how dumb do you think I am?!"
                    else:
                        content += " " + operand + " " + str(value.value) + " = " + str(acc)

        # sends completed message and ends
        bot_handler.send_reply(message, content)
        return;

handler_class = Randy
